rfd/00XX-agent-census.md

```diff
@@ -0,0 +1,237 @@
+
+- [Dockerfile](https://github.com/gravitational/teleport/blob/6f9ad9553a5b5946f57cb35411c598754d3f926b/build.assets/charts/Dockerfile): `ENV TELEPORT_INSTALL_METHOD_DOCKERFILE=true` will be added to the Dockerfile.
+- [`teleport-kube-agent`](https://goteleport.com/docs/reference/helm-reference/teleport-kube-agent) Helm chart: `TELEPORT_INSTALL_METHOD_HELM_KUBE_AGENT` will be set to `true` in the [deployment spec](https://github.com/gravitational/teleport/blob/6f9ad9553a5b5946f57cb35411c598754d3f926b/examples/chart/teleport-kube-agent/templates/deployment.yaml#L129).
+- [`*-ad*.ps1`](https://github.com/gravitational/teleport/tree/6f9ad9553a5b5946f57cb35411c598754d3f926b/lib/web/scripts/desktop): `setx TELEPORT_INSTALL_METHOD_WINDOWS_SCRIPT=true` will be added to one of these scripts as they are the recommended way to configure windows desktops.
+- [`install-node.sh`](https://github.com/gravitational/teleport/blob/6f9ad9553a5b5946f57cb35411c598754d3f926b/lib/web/scripts/node-join/install.sh): `export TELEPORT_INSTALL_METHOD_NODE_SCRIPT="true"` will be added to this script. It is the recommended way to install SSH nodes, apps and many databases. Even though `export` doesn't persist across restarts, we can have the agent persist such value (and maybe all of the values sent in `UpstreamInventoryHello`) when it first starts.
+
+The installation methods that follow won't be tracked for now.
+Later on, we may try to track these if, once we start tracking the above installation methods, we notice that we're not yet covering most methods.
+- tarball: We can add `export TELEPORT_INSTALL_METHOD_TARBALL="true"` to the [`install`](https://github.com/gravitational/teleport/blob/6f9ad9553a5b5946f57cb35411c598754d3f926b/build.assets/install) script. (However, if the customer does not use the `install` script and instead moves the binaries manually, we won't be able to track this installation method.)
+- `.deb`/`.rpm`/`.pkg` packages, APT or YUM repository: It's unclear ATM how these can be tracked.
+- _built from source_: While it's technically possible for customers to build Teleport from source, we won't try to track this installation method as it seems an unlikely use-case.
+- `homebrew`: It's also possible to install Teleport on macOS using `homebrew`. The Teleport package in `homebrew` is not maintained by us, so we will also not track this installation method.
+
+In summary, for now we'll have the following new environment variables and respective value in `UpstreamInventoryHello.InstallMethods`:
+- `TELEPORT_INSTALL_METHOD_DOCKERFILE`: `dockerfile`
+- `TELEPORT_INSTALL_METHOD_HELM_KUBE_AGENT`: `helm-kube-agent`
+- `TELEPORT_INSTALL_METHOD_WINDOWS_SCRIPT`: `windows-script`
+- `TELEPORT_INSTALL_METHOD_NODE_SCRIPT`: `node-script`
+
+##### 8. Container runtime
+
+To determine if the agent is running on Docker, we can first check if the file `/proc/self/cgroup` exists, and if so, if it contains the string `"docker"`.
```

Comment:
Does this method still work on recent Docker/kernel versions? I'm getting this:
```
$ docker run --rm -it ubuntu:22.04 cat /proc/self/cgroup
0::/
$ docker version                                        
Client:
 Version:           20.10.21-ce
 API version:       1.41
 Go version:        go1.18.9
 Git commit:        3056208812eb
 Built:             Tue Dec  6 00:00:00 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.21-ce
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.9
  Git commit:       3056208812eb
  Built:            Tue Dec  6 00:00:00 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.6.12
  GitCommit:        a05d175400b1145e5e6a735a6710579d181e7fb0
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-ga916309fff0f
 docker-init:
  Version:          0.1.7_catatonit
  GitCommit:
```

v20.10.17 and podman 4.3.1 have a similar result, all on my opensuse box. The `/.dockerenv` file does exist, though, so maybe we could also check for that? podman also sets the `container=podman` env var and creates `/run/.containerenv`. This post seems to recommend checking for the env files: https://github.com/containers/podman/issues/3586#issuecomment-661918679

On an older ubuntu box (20.04, docker 20.10.5)  it does work as expected.

Reply:
Changed the suggestion to `/.dockerenv` in 062d39e866e02f59193e0ed38c2a41a42f6201ff.